app/model/TpsPaymentItem.scala

```diff
@@ -19,20 +19,57 @@ package model
 import java.time.LocalDateTime
 
 import model.pcipal.ChargeRefNotificationPcipalRequest
-import play.api.libs.json.{Json, OFormat}
+import play.api.libs.json._
+import play.api.libs.functional.syntax._
 
 case class TpsPaymentItem(
-    paymentItemId:       Option[PaymentItemId]                      = None,
+    paymentItemId:       Option[PaymentItemId],
     amount:              BigDecimal,
     headOfDutyIndicator: HeadOfDutyIndicator,
-    reference:           String,
     updated:             LocalDateTime                              = LocalDateTime.now(),
-    period:              Option[Int],
     customerName:        String,
     chargeReference:     String                                     = "",
-    pcipalData:          Option[ChargeRefNotificationPcipalRequest] = None)
+    pcipalData:          Option[ChargeRefNotificationPcipalRequest] = None,
+    paymentSpecificData: PaymentSpecificData)
 
 object TpsPaymentItem {
-  implicit val format: OFormat[TpsPaymentItem] = Json.format[TpsPaymentItem]
+
+  implicit val writes: OWrites[TpsPaymentItem] = {
+    OWrites[TpsPaymentItem](tpsPaymentItem =>
+      Json.obj(
+        s"amount" -> tpsPaymentItem.amount,
+        s"headOfDutyIndicator" -> tpsPaymentItem.headOfDutyIndicator,
+        s"updated" -> tpsPaymentItem.updated,
+        s"customerName" -> tpsPaymentItem.customerName,
+        s"chargeReference" -> tpsPaymentItem.chargeReference,
+        s"paymentSpecificData" -> tpsPaymentItem.paymentSpecificData
+      )
+        ++ tpsPaymentItem.pcipalData.map(pd => Json.obj("pcipalData" -> pd)).getOrElse(Json.obj())
+        ++ tpsPaymentItem.paymentItemId.map(pid => Json.obj("paymentItemId" -> pid)).getOrElse(Json.obj())
+    )
+  }
+
+  private def typedReads[Psd <: PaymentSpecificData: Reads]: Reads[TpsPaymentItem] = (
+    (__ \ "paymentItemId").readNullable[PaymentItemId] and
+    (__ \ "amount").read[BigDecimal] and
+    (__ \ "headOfDutyIndicator").read[HeadOfDutyIndicator] and
+    (__ \ "updated").read[LocalDateTime] and
+    (__ \ "customerName").read[String] and
+    (__ \ "chargeReference").read[String] and
+    (__ \ "pcipalData").readNullable[ChargeRefNotificationPcipalRequest] and
+    (__ \ "paymentSpecificData").read[Psd]
+
+  )((pid, amnt, hod, updt, cn, cr, pd, psd) => TpsPaymentItem(pid, amnt, hod, updt, cn, cr, pd, psd))
+
+  private val readHod: Reads[TpsPaymentItem] = for {
+    headOfDutyIndicator <- (__ \ "headOfDutyIndicator").read[HeadOfDutyIndicator]
+    payment <- headOfDutyIndicator match {
+      case HeadOfDutyIndicators.B => typedReads[PaymentSpecificDataP800]
```

Comment:
this is non-exhaustive pattern match. Compiler should throw compilation error. Probably `X-fatal-warn` is not set or wart remover is not configured for this.

Reply:
They are both on.  At the moment head of duty indicators only has a B ... for p800.

Reply:
oh, there is only one HOD :) In recon we have plenty of them.